src/nupic/ntypes/Sdr.hpp

```diff
@@ -0,0 +1,778 @@
+     * can be modified without affecting each other.
+     *
+     * @param value An SDR to copy the value of.
+     */
+    void setSDR( const SparseDistributedRepresentation &value ) {
+        NTA_ASSERT( value.dimensions == dimensions );
+        clear();
+
+        dense_valid = value.dense_valid;
+        if( dense_valid ) {
+            dense.assign( value.dense.begin(), value.dense.end() );
+        }
+        flatSparse_valid = value.flatSparse_valid;
+        if( flatSparse_valid ) {
+            flatSparse.assign( value.flatSparse.begin(), value.flatSparse.end() );
+        }
+        sparse_valid = value.sparse_valid;
+        if( sparse_valid ) {
+            for(UInt dim = 0; dim < dimensions.size(); dim++)
+                sparse[dim].assign( value.sparse[dim].begin(), value.sparse[dim].end() );
+        }
+    };
+
+    /**
+     * Calculates the number of true / non-zero values in the SDR.
+     *
+     * @returns The number of true values in the SDR.
+     */
+    UInt getSum()
+        { return getFlatSparse().size(); };
+
+    /**
+     * Calculates the sparsity of the SDR, which is the fraction of bits which
+     * are true out of the total number of bits in the SDR.
+     * I.E.  sparsity = sdr.getSum() / sdr.size
+     *
+     * @returns The fraction of values in the SDR which are true.
+     */
+    Real getSparsity()
+        { return (Real) getSum() / size; };
+
+    /**
+     * Calculates the number of true bits which both SDRs have in common.
+     *
+     * @param sdr, An SDR to compare with, both SDRs must have the same
+     * dimensons.
+     *
+     * @returns Integer, the number of true values which both SDRs have in
+     * common.
+     */
+    UInt overlap(SparseDistributedRepresentation &sdr) {
+        NTA_ASSERT( dimensions.size() == sdr.dimensions.size() );
+        for( UInt i = 0; i < dimensions.size(); i++ )
+            NTA_ASSERT( dimensions[i] == sdr.dimensions[i] );
+
+        UInt ovlp = 0;
+        auto a = this->getDense();
+        auto b = sdr.getDense();
+        for( UInt i = 0; i < size; i++ )
+            ovlp += a[i] && b[i];
+        return ovlp;
+    };
+
+    /**
+     * Make a random SDR, overwriting the current value of the SDR.  The
+     * result has uniformly random activations.
+     *
+     * @param sparsity The sparsity of the randomly generated SDR.
+     *
+     * Both of the following parameters are optional, if neither is given then
+     * the seed 0 is used.
+     * @param seed The seed for the random number generator.
+     * @param rng The random number generator to draw from.
+     */
+    void randomize(Real sparsity) {
+        Random rng( 0 );
+        randomize( sparsity, rng );
+    };
+
+    void randomize(Real sparsity, Random &rng) {
+        NTA_ASSERT( sparsity >= 0. and sparsity <= 1. );
+        UInt nbits = size * sparsity + .5;
+
+        SDR_flatSparse_t range( size );
+        std::iota( range.begin(), range.end(), 0 );
```

Comment:
https://travis-ci.org/htm-community/nupic.cpp/builds/459291239#L3497 

Build fails, just needs to include <numeric> header